### ISSUE TITLE
html5: UI does not allow to do anything if incorrect microphone is chosen automatically

### ISSUE THREAD

**Issue by mikhailnov** [bigbluebutton/bigbluebutton#13503]
BBB v2.4

There may be situations when no micrphone is available or when there is a microphone but it is completely silent.

Let's try on a system aithout any microphones connected:
![image](https://user-images.githubusercontent.com/15802528/137627587-9a365460-63c8-46da-b679-3cdb1c310aa4.png)
![image](https://user-images.githubusercontent.com/15802528/137627594-ff9fb96c-573f-4ddb-ac5d-09e253be7ce9.png)

As you see, we get an error but cannot open a dialog to choose a correct microphone. Chromium browser starts an audio stream, momentally gets an error from BigBlueButton and stops it, so itis not possible to choose audio input device neither via Chromium UI or e.g. via pavucontrol.

**Comment by antobinary** [comment]
I believe this is substantially better on BBB 2.6 default configuration but I am not sure how to reproduce it to verify. Is it an option to try this on test26.bigbluebutton.org ?

**Comment by mikhailnov** [comment]
The "Back" button returns to the screen with buttons to choose a microphone or listen-only mode:
![image](https://user-images.githubusercontent.com/15802528/179032135-2dd4640a-b78d-477c-bdf0-3b59385e227f.png)
![image](https://user-images.githubusercontent.com/15802528/179032014-c2d9ffc6-e713-4a38-8edf-17bffe629fc7.png)

If I remember correctly, I meant that it should show a screen where it is possible another microphone.

I don't know how to reproduce it on real hardware, but I met cases when people had a Windows machine with a working microphone, but for some reasons got that error. And the problem is that they could not get into the screen where they coould configure their microphone. I mean this screen:

![image](https://user-images.githubusercontent.com/15802528/179032634-62e8eb44-2690-4636-8ec8-5484a0a73339.png)

In other words, BigBlueButton sometimes makes in incorrect assumption that there is no microphone available.

**Comment by prlanzarin** [comment]
> If I remember correctly, I meant that it should show a screen where it is possible another microphone.

That's not the case.The `Back` button returns to the initial audio modal, and that's expected behavior (since 2.2).
The audio settings modal you're mentioning is shown first thing after you select `Microphone` in BBB >= 2.6, _or_ clicking the `No/thumbs` down option in the echo test in <= 2.5 - if (and only if, for 2.6 or 2.5) there are valid input devices.

> In other words, BigBlueButton sometimes makes in incorrect assumption that there is no microphone available.

We generally do _not_ assume that kind of stuff, but instead surface what the browser's APIs shows us. If getUserMedia throws an error because it could not find any device, then we can't do much about it because _the browser_ couldn't find an appropriate input device. That doesn't mean there isn't a virtual microphone set up in the OS or that there isn't a HW device plugged in, but it does mean that the browser (we) _have access to neither_. Consequently, we can't redirect the user to a microphone selection screen because _there will be no device_ for the user to choose from.

What we can do is improve the message shown in the error modal. It always suggests it _might_ be a permission issue, but it can be conditional to the error being thrown. We can tailor that for scenarios where there are no input devices to choose (`NotFoundError`).

If you want to further details, please attach the console logs from test26 as well. There's a log line that says the root cause of the error (`Audio settings gUM failed: <something>`).

**Comment by mikhailnov** [comment]
sip.js writes about "requested device not found" into the console:

![image](https://user-images.githubusercontent.com/15802528/179034287-baec168d-b35e-4a01-90c7-12a9b8dec9fc.png)

The following code from sip.js produces it:
https://github.com/onsip/SIP.js/blob/0.20.0/dist/sip-0.20.0.js#L9750

Why does it try to get a description of a device when there are no devices?

Here is s screenshot of chrome://webrtc-internals inside that VM without a microphone:
![image](https://user-images.githubusercontent.com/15802528/179039652-a1f4aeba-4c16-4a23-b4d1-a976b59ea3b5.png)

I downloaded source code of chromium-103.0.5060.53 and found that this message is emitted in third_party/blink/renderer/modules/mediastream/user_media_processor.cc:
https://github.com/chromium/chromium/blob/103.0.5060.53/third_party/blink/renderer/modules/mediastream/user_media_processor.cc#L1739
```cpp
    case MediaStreamRequestResult::NO_HARDWARE:
      user_media_request->Fail(UserMediaRequest::Error::kDevicesNotFound,
                               "Requested device not found");
```
Seems that it looks for devices with reported capture capability:
https://github.com/chromium/chromium/blob/103.0.5060.53/third_party/blink/renderer/modules/mediastream/user_media_processor.cc#L710

Device capabilities seem to be set here: https://github.com/chromium/chromium/blob/103.0.5060.53/third_party/blink/renderer/modules/mediastream/media_devices.cc#L658

I have not found how Chromium decides if a device is able to capture sound or not, maybe it thinks that an available microphone cannot capture sound for some reasons. If it is true, it would be better to show an error saying not that there are no available devices... It seems to be a quite deep issue.

**Comment by mikhailnov** [comment]
> If you want to further details, please attach the console logs from test26 as well. There's a log line that says the root cause of the error (`Audio settings gUM failed: <something>`).

I don't see such a message. But it does not make sense to dig into it when we do ot know how to reproduce a real life issue when there is a microphone, but such an error occurs.

> It always suggests it might be a permission issue

PERMISSION_DENIED and NO_HARDWARE are different cases (https://github.com/chromium/chromium/blob/103.0.5060.53/third_party/blink/renderer/modules/mediastream/user_media_processor.cc#L1722)
I have disallowed microphone on test.bigbluebutton.org and got another error:
![image](https://user-images.githubusercontent.com/15802528/179048518-14bd0039-a99e-4f88-8928-abacef992443.png)
The error from sip.js was another, but the text from BigBlueButton was the same. Maybe BBB can show a different text in this case?

**Comment by prlanzarin** [comment]
> The following code from sip.js produces it:
> https://github.com/onsip/SIP.js/blob/0.20.0/dist/sip-0.20.0.js#L9750
> Why does it try to get a description of a device when there are no devices?

FYI: that's the incorrect SIP.js bundle + version. The correct one is versioned in this repo (based on 0.17.1).
And you're reading the log too literally. It's failing on the gUM phase (which it does _before_ actually generating a description/calling createOffer). tldr it's what I said is happening on my previosu comment.

> I don't see such a message. But it does not make sense to dig into it when we do ot know how to reproduce a real life issue when there is a microphone, but such an error occurs.

I'd be interested in full console logs anyways if you want to attach them.

>  The error from sip.js was another, but the text from BigBlueButton was the same. Maybe BBB can show a different text in this case?

That's what I suggested.

--

Other than that, besides improving the error message (which I'll do), this is out of our hands.
If there's any issue related to this, it would be related to #14429 at this point (which is an actual bug on our side).
As for the fact that Chromium or browser <X> is not picking up input devices is something that should be forwarded to their trackers if you think it's an actual browser issue.

**Comment by prlanzarin** [comment]
> I'd be interested in full console logs anyways if you want to attach them.

Bah! I just verified that test26 is not up-to-date with v2.6.x-release so that's why the audio join flow is still the alpha.1/old one. So no need to forward full logs - the previous screenshot suffices, thanks.

**Comment by mikhailnov** [comment]
> I'd be interested in full console logs anyways if you want to attach them.
https://clbin.com/JyCdc
> So no need to forward full logs
ok ) but have already done that, feel free to ask me to upload new logs if needed

> If there's any issue related to this, it would be related to https://github.com/bigbluebutton/bigbluebutton/issues/14429 at this point (which is an actual bug on our side).
Thanks for pointing to it

**Comment by Scroody** [comment]
I was also able to reproduce this behaviour: I work using a laptop, however I use a normal monitor, the laptop screen is too small, and because of that I leave my laptop closed, and that caused the speakers and mic do be unavailable, causing the same error. So if that ever happens to someone that uses the same setup, consider opening the laptop.
